### DISCUSSION TITLE
`linq2db/ASP.NET` DI in DataConnection problem

### DISCUSSION THREAD

**Discussion by hyperion-cs** [linq2db/linq2db#3632] [Q&A] [answered]
Friends, hello.

The wonderful **linq2db** [supports](https://linq2db.github.io/articles/get-started/asp-dotnet-core/index.html) working inside ASP.NET, including dependency injection.
However, I faced the problem of dependency injection into a derived class from `DataConnection`.
Namely, at the moment there is a hard constraint for the class constructor to look something like this:
```cs
public FooDataConnection(LinqToDbConnectionOptions<FooDataConnection> options) : base(options) { ... }
```

So if, for example, we go in to inject something into this class through the constructor:
```cs
private readonly ISomeService _iSomeService;
...
public FooDataConnection(ISomeService iSomeService, LinqToDbConnectionOptions<FooDataConnection> options) : base(options)
{
	...
	_iSomeService = iSomeService;
	...
}
```

Then we get an error of the following form:
```
System.ArgumentException: 'Missing constructor accepting 'LinqToDBConnectionOptions' on type FooDataConnection'
```

This seems to be wrong behavior. Because the constructor accepting `LinqToDbConnectionOptions` actually exists.
Does anyone have any idea how to fix this?

**Comment by viceroypenguin** [comment]
Would it be possible for you to put together a demo project with a minimal code to replicate this issue? There's no indication here what the DI system is trying to construct when throwing this exception, so some more context would be helpful.

**Reply by hyperion-cs** [reply]
@viceroypenguin, of course. [Here](https://github.com/hyperion-cs/linq2db-3632-problem) you can find a repository with a minimal demo project that reproduces the problem.

**Reply by viceroypenguin** [reply]
Thanks! I'll take a look soon.

**Reply by viceroypenguin** [reply]
Currently, the automatic `.AddLinqToDBContext()` only supports a constructor that receives a single parameter of type `LinqToDBConnectionOptions`. 

However, I don't see any reason it couldn't support constructors that receive other parameters as long as it only receives one `LinqToDBConnectionOptions` parameter. I'll create an issue for it and we'll get it into 4.2.0.

**Reply by hyperion-cs** [reply]
@viceroypenguin, thank you very much. We will wait for a fix.

**Reply by viceroypenguin** [reply]
This has been fixed and will be released in 4.1.1

**Reply by hyperion-cs** [reply]
@viceroypenguin, great work, you are amazing!

**Comment by hyperion-cs** [comment, marked as answer]
The issue is resolved in https://github.com/linq2db/linq2db/discussions/3632#discussioncomment-3055759, https://github.com/linq2db/linq2db/pull/3634.
